listenerutil/parse.go

```diff
@@ -387,3 +413,142 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
 		return "", fmt.Errorf("multiple addresses found (%q), please configure one", out)
 	}
 }
+
+const strictTransportSecurity = "max-age=31536000; includeSubDomains"
+const xContentTypeOptions = "nosniff"
+const cacheControl = "no-store"
+const apiContentSecurityPolicy = "default-src 'none'"
+const uiContentSecurityPolicy = "default-src 'none'; script-src 'self'; frame-src 'self'; font-src 'self'; connect-src 'self'; img-src 'self' data:*; style-src 'self'; media-src 'self'; manifest-src 'self'; style-src-attr 'self'; frame-ancestors 'self'"
+
+// parseCustomResponseHeaders takes raw config values for the "custom_ui_response_headers"
+// and "custom_api_response_headers". It makes sure the config entry is passed in as a map
+// of status code to a map of header name and header values. It verifies the validity of the
+// status codes, and header values. It also adds the default headers values for "Cache-Control",
+// "Strict-Transport-Security", "X-Content-Type-Options", and "Content-Security-Policy".
+func parseCustomResponseHeaders(responseHeaders interface{}, uiHeaders bool) (map[int]map[string][]string, error) {
+	h := make(map[int]map[string][]string)
+	// if responseHeaders is nil, we still should set the default custom headers
+	if responseHeaders == nil {
+		h[0] = map[string][]string{
+			"Strict-Transport-Security": {strictTransportSecurity},
+			"X-Content-Type-Options":    {xContentTypeOptions},
+			"Cache-Control":             {cacheControl},
+		}
+		if uiHeaders {
+			h[0]["Content-Security-Policy"] = []string{uiContentSecurityPolicy}
+		} else {
+			h[0]["Content-Security-Policy"] = []string{apiContentSecurityPolicy}
+		}
+		return h, nil
+	}
+
+	customResponseHeader, ok := responseHeaders.([]map[string]interface{})
```

Comment:
Rather than doing these checks here for `[]map[string]interface{}` consider using `mapstructure` to do the decoding for you. Not that this is wrong, but it saves some boilerplate. Probably whether it's worth it depends on if we'd want to allow weak decoding that mapstructure supports so you can e.g. type either `200` or `"200"`. Pros and cons to that...

Reply:
would there be any benefits to the hcl by implementing `mapstructure`? I would personally prefer to avoid using an external package unless there are explicit benefits to using it

Reply:
As I said, depends on if we want to allow weak decoding, and also handles a lot of type checking/reflect stuff for us...makes it easy to declare the structure you want and have it populated for you. `mapstructure` is used all over HC code (it's our lib) so if it's helpful it's a totally fine dep to bring in.

Reply:
so I've been reading through the [mapstructure docs](https://pkg.go.dev/github.com/mitchellh/mapstructure), and it seems really helpful for decoding known structures, but I'm not sure how it would help here, since we need to do additional parsing on all the values anyway via `convertStatusCode` and `parseHeaders`

Reply:
Parsing is separate from validation. Using mapstructure would allow people to specify status codes as either strings or ints. Right now you're forcing headers to be specified as strings, which may not be intuitive since HCL supports ints. I'm fine with it in this case because we also need to support 2xx style values, but just an FYI.